### PR TITLE
Adopt work interval API

### PR DESCRIPTION
**Pull request by pvollan**
#### f31d0edc7b5e836fb7570ba0b4a2e28a0d092de2
<pre>
Adopt work interval API
<a href="https://bugs.webkit.org/show_bug.cgi?id=246829">https://bugs.webkit.org/show_bug.cgi?id=246829</a>
rdar://101398504

Reviewed by NOBODY (OOPS!).

Adopt work interval API on macOS and iOS.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/cocoa/WorkgroupInterval.cpp: Added.
(WTF::WorkgroupInterval::WorkgroupInterval):
(WTF::WorkgroupInterval::createInterval):
(WTF::WorkgroupInterval::join):
(WTF::WorkgroupInterval::leave):
(WTF::WorkgroupInterval::start):
(WTF::WorkgroupInterval::update):
(WTF::WorkgroupInterval::finish):
* Source/WTF/wtf/cocoa/WorkgroupInterval.h: Added.
(WTF::WorkgroupIntervalScope::WorkgroupIntervalScope):
(WTF::WorkgroupIntervalScope::~WorkgroupIntervalScope):
* Source/WebCore/page/Page.cpp:
(WebCore::workgroupRendering):
(WebCore::Page::~Page):
(WebCore::Page::updateRendering):
(WebCore::Page::setIsVisuallyIdleInternal):
* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
</pre>
https://github.com/WebKit/WebKit/commit/f31d0edc7b5e836fb7570ba0b4a2e28a0d092de2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93808 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3003 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24390 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103446 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163771 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97801 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3019 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31246 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86136 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99457 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99470 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/76/builds/3019 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80240 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29173 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/76/builds/3019 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/43/builds/24390 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72128 "Passed tests") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/85063 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37641 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/43/builds/24390 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80193 "Built successfully") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35499 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/43/builds/24390 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27663 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39377 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/80240 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82837 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41311 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/43/builds/24390 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18721 "Passed tests") |

### CODE REVIEW

File: Source/WebCore/page/Page.cpp [hunk 1]
```diff
@@ -435,6 +443,11 @@ Page::~Page()
     m_pluginInfoProvider->removePage(*this);
     m_userContentProvider->removePage(*this);
     m_visitedLinkStore->removePage(*this);
+    
+#if PLATFORM(COCOA)
+    if (m_intervalStarted)
+        workgroupRendering().finish();
+#endif
```
Comment (geoffreygaren) [inline Source/WebCore/page/Page.cpp]: Because the workgroup interval is scoped to Page::updateRendering, it's impossible in a well-formed program for ~Page() to run while we have a live interval -- that would mean that we had deleted 'this' in the midst of a member function. So, I think you can remove this.

File: Source/WebCore/page/Page.cpp [hunk 2]
```diff
@@ -2064,6 +2086,13 @@ void Page::setIsVisuallyIdleInternal(bool isVisuallyIdle)
     if (isVisuallyIdle == m_throttlingReasons.contains(ThrottlingReason::VisuallyIdle))
         return;
 
+#if PLATFORM(COCOA)
+    if (isVisuallyIdle && m_intervalStarted) {
+        workgroupRendering().finish();
+        m_intervalStarted = false;
+    }
+#endif
+
```
Comment (geoffreygaren) [inline Source/WebCore/page/Page.cpp]: Because the workgroup interval is scoped to Page::updateRendering, it's impossible in a well-formed program for a visually idle notification to be processed while we have a live interval -- since notifications are only processed when we return to the run loop. So, I think you can remove this.

File: Source/WebCore/page/Page.h [hunk 3]
```diff
@@ -1348,6 +1352,11 @@ class Page : public Supplementable<Page>, public CanMakeWeakPtr<Page> {
 #endif
 
     ContentSecurityPolicyModeForExtension m_contentSecurityPolicyModeForExtension { ContentSecurityPolicyModeForExtension::None };
+    
+#if PLATFORM(COCOA)
+    bool m_joined { false };
+    bool m_intervalStarted { false };
```
Comment (geoffreygaren) [inline Source/WebCore/page/Page.h]: If I'm right above, then you can remove m_intervalStarted.

File: Source/WebCore/page/Page.h [hunk 4]
```diff
@@ -1348,6 +1352,11 @@ class Page : public Supplementable<Page>, public CanMakeWeakPtr<Page> {
 #endif
 
     ContentSecurityPolicyModeForExtension m_contentSecurityPolicyModeForExtension { ContentSecurityPolicyModeForExtension::None };
+    
+#if PLATFORM(COCOA)
+    bool m_joined { false };
```
Comment (geoffreygaren) [inline Source/WebCore/page/Page.h]: I don't think you need this because you're using call_once instead.

File: Source/WTF/wtf/cocoa/WorkgroupInterval.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    auto rv = work_interval_leave();
+    WTFLogAlways("Workgroup leave %d", rv);
+    return !rv;
+}
+
+bool WorkgroupInterval::start(uint64_t startOffsetNanoSeconds, uint64_t deadlineNanoSeconds)
+{
+    Locker locker { m_lock };
+    if (!m_instance)
+        m_instance = work_interval_instance_alloc(m_interval);
```
Comment (geoffreygaren) [inline Source/WTF/wtf/cocoa/WorkgroupInterval.cpp]: Should we just allocate this in our constructor?

File: Source/WTF/wtf/cocoa/WorkgroupInterval.cpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WorkgroupInterval.h"
+
+namespace WTF {
+
+WorkgroupInterval::WorkgroupInterval(const char* name)
+    : m_interval(createInterval(name))
```
Comment (geoffreygaren) [inline Source/WTF/wtf/cocoa/WorkgroupInterval.cpp]: Should we deallocate this in our destructor?

File: Source/WTF/wtf/cocoa/WorkgroupInterval.h [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    WTF_EXPORT_PRIVATE bool update(uint64_t deadlineNanoSeconds);
+    WTF_EXPORT_PRIVATE bool finish();
+
+private:
+    WTF_EXPORT_PRIVATE static work_interval_t createInterval(const char* name);
+
+    work_interval_t m_interval { nullptr };
+    work_interval_instance_t m_instance { nullptr };
+    std::optional<std::thread::id> m_startThread;
+    Lock m_lock;
```
Comment (geoffreygaren) [inline Source/WTF/wtf/cocoa/WorkgroupInterval.h]: I don't think we're using the multi-threading feature anymore.

**Review by geoffreygaren — CHANGES_REQUESTED** [review]

File: Source/WTF/wtf/cocoa/WorkgroupInterval.h (hunk 7 above)
Comment (pvollan) [reply Source/WTF/wtf/cocoa/WorkgroupInterval.h]: Thanks! The latest patch should address these issues.

File: Source/WTF/wtf/cocoa/WorkgroupInterval.cpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    work_interval_instance_free(m_instance);
+}
+
+work_interval_t WorkgroupInterval::createInterval(const char*)
+{
+    work_interval_t workIntervalHandle = nullptr;
+    const int rv = work_interval_create(&workIntervalHandle, WORK_INTERVAL_TYPE_CA_CLIENT | WORK_INTERVAL_FLAG_JOINABLE);
+    RELEASE_ASSERT(!rv);
+    RELEASE_ASSERT(workIntervalHandle);
+    WTFLogAlways("Workgroup interval = %p", workIntervalHandle);
```
Comment (smfr) [inline Source/WTF/wtf/cocoa/WorkgroupInterval.cpp]: Need to remove the logging.

File: Source/WTF/wtf/cocoa/WorkgroupInterval.cpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+bool WorkgroupInterval::leave()
+{
+    auto rv = work_interval_leave();
+    WTFLogAlways("Workgroup leave %d", rv);
+    return !rv;
+}
+
+bool WorkgroupInterval::start(uint64_t startOffsetNanoSeconds, uint64_t deadlineNanoSeconds)
```
Comment (smfr) [inline Source/WTF/wtf/cocoa/WorkgroupInterval.cpp]: I think this should just take a `Seconds` duration. We always call it to start "now".

File: Source/WTF/wtf/cocoa/WorkgroupInterval.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+{
+    if (m_startThread)
+        return false;
+    mach_timebase_info_data_t timebase;
+    mach_timebase_info(&timebase);
```
Comment (smfr) [inline Source/WTF/wtf/cocoa/WorkgroupInterval.cpp]: Would be nicer to use `MonotonicTime::toMachAbsoluteTime` math, rather than replicating that math here.

File: Source/WTF/wtf/cocoa/WorkgroupInterval.cpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    }
+    m_startThread = std::this_thread::get_id();
+    return true;
+}
+
+bool WorkgroupInterval::update(uint64_t deadlineNanoSeconds)
+{
+    mach_timebase_info_data_t timebase;
+    mach_timebase_info(&timebase);
+    int64_t deadlineAbsolute = (deadlineNanoSeconds * timebase.denom) / timebase.numer;
```
Comment (smfr) [inline Source/WTF/wtf/cocoa/WorkgroupInterval.cpp]: Ditto.

File: Source/WTF/wtf/cocoa/WorkgroupInterval.cpp [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+bool WorkgroupInterval::finish()
+{
+    if (!m_startThread) {
+        WTFLogAlways("Workgroup finish error, start not called");
+        return false;
+    }
+    if (*m_startThread != std::this_thread::get_id()) {
+        WTFLogAlways("Workgroup finish error, start called on different thread");
```
Comment (smfr) [inline Source/WTF/wtf/cocoa/WorkgroupInterval.cpp]: Seems like this should be an ASSERT

File: Source/WTF/wtf/cocoa/WorkgroupInterval.h [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace WTF {
+
+class WorkgroupInterval {
+public:
+    WTF_EXPORT_PRIVATE WorkgroupInterval(const char* name);
+    WTF_EXPORT_PRIVATE ~WorkgroupInterval();
+    WTF_EXPORT_PRIVATE bool join();
+    WTF_EXPORT_PRIVATE bool leave();
+    WTF_EXPORT_PRIVATE bool start(uint64_t startOffsetNanoSeconds, uint64_t deadlineNanoSeconds);
```
Comment (smfr) [inline Source/WTF/wtf/cocoa/WorkgroupInterval.h]: Use Seconds here.

File: Source/WTF/wtf/cocoa/WorkgroupInterval.h [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+    work_interval_t m_interval { nullptr };
+    work_interval_instance_t m_instance { nullptr };
+    std::optional<std::thread::id> m_startThread;
+};
+
+class WorkgroupIntervalScope {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    WorkgroupIntervalScope(WorkgroupInterval& interval, int64_t startOffsetNanoSeconds, uint64_t deadlineNanoSeconds)
```
Comment (smfr) [inline Source/WTF/wtf/cocoa/WorkgroupInterval.h]: Use Seconds here.

File: Source/WebCore/page/Page.cpp [hunk 15]
```diff
@@ -1675,6 +1683,15 @@ void Page::updateRendering()
 
     m_lastRenderingUpdateTimestamp = MonotonicTime::now();
 
+#if HAVE(WORK_INTERVAL_API)
+    static std::once_flag onceFlag;
```
Comment (smfr) [inline Source/WebCore/page/Page.cpp:1687]: This code needs to move somewhere else. The `WorkgroupIntervalScope` should be a one-liner here.

Also consider making a stub for other platforms so that there is no #ifdeffing needed here.

Comment (pvollan) [reply Source/WebCore/page/Page.cpp:1687]: Thanks for reviewing! I will address these issues.

File: Source/WTF/wtf/cocoa/WorkgroupInterval.cpp (hunk 12 above)
Comment (pvollan) [reply Source/WTF/wtf/cocoa/WorkgroupInterval.cpp]: Will fix!

File: Source/WTF/wtf/cocoa/WorkgroupInterval.h (hunk 13 above)
Comment (pvollan) [reply Source/WTF/wtf/cocoa/WorkgroupInterval.h]: Will fix!

File: Source/WTF/wtf/cocoa/WorkgroupInterval.h (hunk 14 above)
Comment (pvollan) [reply Source/WTF/wtf/cocoa/WorkgroupInterval.h]: Will fix.

File: Source/WTF/wtf/cocoa/WorkgroupInterval.cpp (hunk 5 above)
Comment (pvollan) [reply Source/WTF/wtf/cocoa/WorkgroupInterval.cpp]: Fixed!

File: Source/WebCore/page/Page.h (hunk 4 above)
Comment (pvollan) [reply Source/WebCore/page/Page.h]: Fixed!

File: Source/WebCore/page/Page.h [hunk 16]
```diff
@@ -82,6 +82,10 @@
 #include "DeviceOrientationUpdateProvider.h"
 #endif
 
+#if PLATFORM(COCOA)
+#include <wtf/cocoa/WorkgroupInterval.h>
```
Comment (smfr) [inline Source/WebCore/page/Page.h:86]: Why is this needed in Page.h?

File: Source/WebCore/page/Page.cpp [hunk 17]
```diff
@@ -262,6 +262,14 @@ static constexpr OptionSet<ActivityState::Flag> pageInitialActivityState()
     return { ActivityState::IsVisible, ActivityState::IsInWindow };
 }
 
+#if HAVE(WORK_INTERVAL_API)
+static WTF::WorkgroupInterval& workgroupRendering()
```
Comment (smfr) [inline Source/WebCore/page/Page.cpp:266]: I'd still prefer to see this outside of Page.cpp.

File: Source/WebCore/page/Page.h (hunk 16 above)
Comment (pvollan) [reply Source/WebCore/page/Page.h:86]: Good catch, this should go in the cpp file. Will fix, this will likely be in a new PR, though.

Thanks for reviewing!

File: Source/WebCore/page/Page.cpp (hunk 17 above)
Comment (pvollan) [reply Source/WebCore/page/Page.cpp:266]: Yes, I agrees. Will fix, likely in a new PR.

Thanks!
